src/nupic/algorithms/Segment.cpp

```diff
@@ -392,7 +389,7 @@ void Segment::freeNSynapses(UInt numToFree,
   }
 
   // Now sort the list of candidate synapses
-  std::sort(candidates.begin(), candidates.end(), InPermanenceOrder());
+  std::stable_sort(candidates.begin(), candidates.end(), InPermanenceOrder());
```

Comment:
Could this benefit from a comment such as: "Use stable sort for the case where we have multiple 
synapses with the same lowest permanence"?